applications/RomApplication/custom_strategies/rom_builder_and_solver.h

```diff
@@ -333,113 +379,107 @@ class ROMBuilderAndSolver : public BuilderAndSolver<TSparseSpace, TDenseSpace, T
+
+                    // clean local elemental memory
+                    pScheme->CleanMemory(*(it_el.base()));
+                }
             }
-        }
 
-        for (int k = 0; k < nconditions; k++){
-            ModelPart::ConditionsContainerType::iterator it = cond_begin + k;
-
-            //detect if the element is active or not. If the user did not make any choice the element
-            //is active by default
-            bool condition_is_active = true;
-            if ((it)->IsDefined(ACTIVE))
-                condition_is_active = (it)->Is(ACTIVE);
-
-            if (condition_is_active){
-                Condition::DofsVectorType dofs;
-                it->GetDofList(dofs, CurrentProcessInfo);
-                //calculate elemental contribution
-                pScheme->Condition_CalculateSystemContributions(*(it.base()), LHS_Contribution, RHS_Contribution, EquationId, CurrentProcessInfo);
-
-                //assemble the elemental contribution - here is where the ROM acts
-                //compute the elemental reduction matrix Phi
-                const auto &r_geom = it->GetGeometry();
-                Matrix PhiElemental(r_geom.size() * mNodalDofs, mRomDofs);
-
-                for (unsigned int i = 0; i < r_geom.size(); ++i){
-                    const Matrix &rom_nodal_basis = r_geom[i].GetValue(ROM_BASIS);
-                    for (unsigned int k = 0; k < rom_nodal_basis.size1(); ++k){
-                        if (dofs[i * mNodalDofs + k]->IsFixed())
-                            row(PhiElemental, i * mNodalDofs + k) = ZeroVector(PhiElemental.size2());
-                        else
-                            row(PhiElemental, i * mNodalDofs + k) = row(rom_nodal_basis, k);
-                    }
+            #pragma omp for nowait
+            for (int k = 0; k < nconditions; k++){
+                ModelPart::ConditionsContainerType::iterator it = cond_begin + k;
+
+                //detect if the element is active or not. If the user did not make any choice the condition
+                //is active by default
+                bool condition_is_active = true;
+                if ((it)->IsDefined(ACTIVE))
+                    condition_is_active = (it)->Is(ACTIVE);
+                if (condition_is_active){
+                    Condition::DofsVectorType dofs;
+                    it->GetDofList(dofs, CurrentProcessInfo);
+                    //calculate elemental contribution
+                    pScheme->Condition_CalculateSystemContributions(*(it.base()), LHS_Contribution, RHS_Contribution, EquationId, CurrentProcessInfo);
+                    const auto &geom = it->GetGeometry();
+                    if(PhiElemental.size1() != dofs.size() || PhiElemental.size2() != mRomDofs)
+                        PhiElemental.resize(dofs.size(), mRomDofs,false);
+                    GetPhiElemental(PhiElemental, dofs, geom);
+                    Matrix aux = prod(LHS_Contribution, PhiElemental);

```

Comment:
here you could reuse the same matrix as in the previous loop (resizing it as needed)

Reply:
done